query.go

```diff
@@ -1077,9 +1106,27 @@ func (s *ChainService) sendTransaction(tx *wire.MsgTx, options ...QueryOption) e
 							tx, nil, qo.encoding,
 						)
 
-						numReplied++
-
-						close(peerQuit)
+						// Peers might send the INV
+						// request multiple times, we
+						// need to de-duplicate them
+						// using a map.
+						replyChan := make(chan struct{})
+						replies[sp.ID()] = replyChan
+
+						// Okay, so the peer responded
+						// with an INV message, and we
+						// sent out the TX. If we never
+						// hear anything back from the
+						// peer it means they accepted
+						// the tx. If we get a reject,
+						// things are clear as well. But
+						// what if they are just slow to
+						// respond? We'll give them
+						// another bit of time to
+						// respond.
+						closeEventually(
+							peerQuit, replyChan,
+						)
```

Comment:
Hmm, sorry if I broke this.  I must have been mistaken in thinking that our `inv` request would be completed after receiving a `getdata` response (this case) and handling it by queuing the msgtx send.

Is the issue that after this first `getdata` response case handles it, that there may be a subsequent `reject` that should be handled by this same closure?  My (apparently bad) assumption was that further responses to the tx data was handled elsewhere since that msgtx send was queued up for asynchronous sending.

Looks like the `QueryRejectTimeout` is meant to handle any such second response following handling of the first `getdata` response?  1 second is certainly better than the full 5 second broadcast timeout that was always hit before!  What you're doing with `replyChan` makes sense in that context.

Thanks for the proper resolution to the broadcast timeout issue I was trying to solve.  :)

Reply:
> Hmm, sorry if I broke this.

No problem. I guess that's what our unit and integration tests in lnd are there for.

The actual problem is that publishing a TX over the p2p network isn't a synchronous process.
What happens is the following:
 1. The Neutrino client sends out an INV message with just the TXID of the so far unpublished TX ("hey y'all, have you ever heard of this TX?")
 2. Every peer will look at their mempool and if they don't find it, they'll answer back with a GetData message ("this is news to me, mind sending over that full TX?")
    a) Now it could be that the peer just doesn't respond in time. In that case we just ignore that peer
    b) It could also be that the peer already knows of the TX (for example we published it before). In that case it won't react at all (I think).
  3. We send the full TX out to each peer that requested it as an MsgTx ("here it is, plz forward!").
  4. The peer now either accepts the TX and forwards it or rejects it.
    a) If the peer accepts, we don't get any ACK or other response. After some time we just need to assume they're okay with it
    b) If the peer rejects, they send back a MsgReject and we know something's up.

From what I understand, all of this happens in the closure above. I'm not 100% sure my solution works in all cases. But at least it should reduce the time for a normal, successful publish from 5 seconds to 1 second.

Why 1 second and not 5 as before?
The assumption here is that if a peer responds quickly (e.g. before the overall timeout), it should respond with a reject reasonably quickly too. If a peer fails to respond within a second, we still have the threshold detection to recognize invalid transactions.

Reply:
> b) If the peer rejects, they send back a MsgReject and we know something's up.

FWIW, the latest `bitcoind` nodes won't send a reject message, as result we can end up in a more murky state where we don't know if they already have it, or think it's invalid.

Reply:
> This last bit is what I figured wasn't handled in this closure but rather whatever is sending the tx, but it does make sense it would be need to be to associate this reject with this inv message.

Yeah it ends up happening in this closure. This essentially hooks into every message received by the peer, with us only acting on certain messages. 

This section of the code is actually using an _older_ instance of this query code. The newer (re-designed) query interface is currently only used in the block manager: https://github.com/lightninglabs/neutrino/blob/master/blockmanager.go#L1052

We have some issues that track the process of moving over to the new interface uniformly.